### PR TITLE
docs: add migration page to user guide

### PR DESCRIPTION
**Pull request by jhamman**
This PR adds a page describing the migration process for users upgrading to zarr 3.0.

Notes for reviewers:

- This was split out from https://github.com/zarr-developers/zarr-python/pull/2568 and builds on https://github.com/zarr-developers/zarr-python/pull/2589. Initially started in #2102.

Closes https://github.com/zarr-developers/zarr-python/issues/1798

### CODE REVIEW

File: docs/user-guide/v3_migration.rst [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+This page provides a guide highlighting the most notable changes to help you
+migrate your code from version 2 to version 3.
+
+Zarr-Python 3 represents a major refactor of the Zarr-Python codebase. Some of the
+goals motivating this refactor included:
+
+* adding support for the Zarr V3 specification (alongside the Zarr V2 specification)
+* cleaning up internal and user facing APIs
+* improving performance (particularly in high latency storage environments like
+  cloud object store)
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:16]:
```suggestion
Zarr-Python 3 represents a major refactor of the Zarr-Python codebase. Some of the
goals motivating this refactor included:

* adding support for the Zarr V3 specification (alongside the Zarr V2 specification)
* cleaning up internal and user facing APIs
* improving performance (particularly in high latency storage environments like
  cloud object store)
  
To accomodate this, Zarr-Python 3.0 introduces a number of changes to the API, including a number
of significant breaking changes and pending deprecations.

This page provides a guide highlighting the most notable changes to help you
migrate your code from version 2 to version 3.
```
I think this is a more logical order, which motivates the changes first.

File: docs/user-guide/v3_migration.rst [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+Compatibility target
+--------------------
+
+The goals described above necessitated some breaking changes to the API (hence the
+major version update), but we have attempted to maintain ~95% backwards compatibility
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:22]:
```suggestion
major version update), but where poissible we have maintained backwards compatibility
```

File: docs/user-guide/v3_migration.rst [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+in the most widely used parts of the API. This in the :class:`zarr.Array` and
+:class:`zarr.Group` classes and the "top-level API" (e.g. :func:`zarr.open_array` and
+:func:`zarr.open_group`).
+
+Getting ready for 3.0
+---------------------
+
+Ahead of the 3.0 release, we suggest projects that depend on Zarr-Python take the
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:30]: If the plan is to release in 6 days time, then I don't think there's much point adding a new section to the docs that says "Ahead of the 3.0 release...". I'd suggest moving the "Using zarr Python v2" section up here, and adding the pinning advice to that section. Then the flow is

1. How to continue using v2
2. How to transition to v3

File: docs/user-guide/v3_migration.rst [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+Migration Guide
+---------------
+
+The following sections provide details on the most important changes in Zarr-Python 3.
+
+The Array class
+~~~~~~~~~~~~~~~
+
+1. Disallow direct construction - use :func:`zarr.open_array` or :func:`zarr.create_array`
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:93]: It's not clear here whether direct contstruction is:

- No longer possible
- Deprecated
- Possible, but discouraged

File: docs/user-guide/v3_migration.rst [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+2. Defaulting to ``zarr_format=3`` - newly created arrays will use the version 3 of the
+   Zarr specification. To continue using version 2, set ``zarr_format=2`` when creating arrays
+   or set ``default_zarr_version=2`` in :ref:`config`.
+
+The Group class
+~~~~~~~~~~~~~~~
+
+1. Disallow direct construction - use :func:`zarr.open_group` or :func:`zarr.create_group`
+   instead of directly constructing the :class:`zarr.Group` class.
+2. Deprecated most of the h5py compatibility methods. The following migration is suggested:
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:105]:
```suggestion
2. Most of the h5py compatibility methods are deprecated. The following migration is suggested:
```

Comment (dstansby) [inline docs/user-guide/v3_migration.rst:105]: Are the suggested functions drop in replacements with the same arguments? If not, can we provide a clearer guide on how to translate existing calls to create_dataset/require_dataset to create_array/require_array?

File: docs/user-guide/v3_migration.rst [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+   - Use :func:`zarr.Group.create_array` in place of :func:`zarr.Group.create_dataset`
+   - Use :func:`zarr.Group.require_array` in place of :func:`zarr.Group.require_dataset`
+
+The Store class
+~~~~~~~~~~~~~~~
+
+Some of the biggest changes in Zarr-Python 3 are found in the ``Store`` class. The most notable changes to the Store API are:
+
+1. Replaced the ``MutableMapping`` base class in favor of a custom abstract base class (:class:`zarr.abc.store.Store`).
+2. Switched to a primarily Async interface.
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:116]: What does this mean in conrete terms for a user? Is it just a behind the scenes change (if so, why was it done? what benefits does it bring?), or does the user have to adapt their code for this change?

File: docs/user-guide/v3_migration.rst [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+Beyond the changes store interface, a number of deprecated stores were also removed in Zarr-Python 3:
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:118]: What are we recommending users who use these stores do? Create their own custom stores? Or migrate to other options?

File: docs/user-guide/v3_migration.rst [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+- ``N5Store``
+- ``DBMStore``
+- ``LMDBStore``
+- ``SQLiteStore``
+- ``MongoDBStore``
+- ``RedisStore``
+- ``ABSStore``
+
+Dependencies Changes
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:128]:
```suggestion
Dependency Changes
```

File: docs/user-guide/v3_migration.rst [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+~~~~~~~~~~~~~~~~~~~~
+
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:130]:
```suggestion

When installing using `pip`:
```
Since this doesn't apply to conda installs.

File: docs/user-guide/v3_migration.rst [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+  ``fsspec``, required for remote data access.
+- The new ``gpu`` dependency group can be used to install a supported version of
+  ``cuda``, required for GPU functionality.
+- The ``jupyter`` optional dependency group has been removed, since v3 contains no
+  jupyter specific functionality.
+
+Configuration
+~~~~~~~~~~~~~
+
+There is a new configuration system based on `donfig <https://github.com/pytroll/donfig>`_,
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:141]: I think this should be in a what's new, not a migration guide, since there's no functionality from v2 in this section.

File: docs/user-guide/v3_migration.rst [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+- Default array order in memory ``array.order``
+- Default codecs ``array.v3_default_codecs`` and ``array.v2_default_compressor``
+- Whether empty chunks are written to storage ``array.write_empty_chunks``
+- Async and threading options, e.g. ``async.concurrency`` and ``threading.max_workers``
+- Selections of implementations of codecs, codec pipelines and buffers
+
+Miscellaneous
+~~~~~~~~~~~~~
+
+- The keyword argument ``zarr_version`` has been deprecated in favor of ``zarr_format``.
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:165]: What functions is this an argument to?

File: docs/user-guide/v3_migration.rst [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+after the 3.0 release:
+
+- The following functions / methods have not been ported to Zarr-Python 3 yet:
+
+  * :func:`zarr.copy`
+  * :func:`zarr.copy_all`
+  * :func:`zarr.copy_store`
+  * :func:`zarr.Group.move`
+
+- The following options in the top-level API have not been ported to Zarr-Python 3 yet.
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:181]: What does "options in the top-level API" mean? Config options? Or arguments (and if so, to which functions?)

File: docs/user-guide/v3_migration.rst [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+
+  * ``cache_attrs``
+  * ``cache_metadata``
+  * ``chunk_store``
+  * ``meta_array``
+  * ``object_codec``
+  * ``synchronizer``
+  * ``dimension_separator``
+
+- The following features have not been ported to Zarr-Python 3 yet:
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:194]: Do we want folks to open github issues if they want these features too?

File: docs/user-guide/v3_migration.rst [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+
+- The following functions / methods have not been ported to Zarr-Python 3 yet:
+
+  * :func:`zarr.copy`
+  * :func:`zarr.copy_all`
+  * :func:`zarr.copy_store`
+  * :func:`zarr.Group.move`
+
+- The following options in the top-level API have not been ported to Zarr-Python 3 yet.
+  If these options are important to you, please open a
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:182]: Perhaps move this request for issues up to the top of this section, so it applies to all the bullet points and not just this one?

File: docs/user-guide/v3_migration.rst [hunk 15]
```diff
@@ -0,0 +1,204 @@
+3.0 Migration Guide
+===================
+
+Zarr-Python 3.0 introduces a number of changes to the API, including a number
+of significant breaking changes and pending deprecations.
+
+This page provides a guide highlighting the most notable changes to help you
+migrate your code from version 2 to version 3.
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:8]:
```suggestion
This page provides a guide explaining the breaking changes to help you
migrate your code from version 2 to version 3.
If we have missed anything, please open a `GitHub issue <https://github.com/zarr-developers/zarr-python/issues/new>`_ so we can impove this guide.
```

**Review by dstansby — CHANGES_REQUESTED** [review]
This is a good start, but I think needs some major improvements and additions. In particular: 

- Removing the "getting ready" section, since there won't be a decent period of time between these docs existing and the final release for folks to prepare for v3.
- Being more explicit about what's change and how to adapt existing code (I left inline comments for this)
- Listing where keyword only arguments have been introduced (e.g., I noticed these in [zarr.create](https://zarr.readthedocs.io/en/latest/_autoapi/zarr/index.html#zarr.create)), as this is a major breaking change where users will have to update code

I also compared the v2 and v3 API, and there's some major additions that need making to this guide so users can migrate:
- Explaining how to adapt code that previously used [zarr.hierarchy](https://zarr.readthedocs.io/en/stable/api/hierarchy.html)
- Explaining how to migrate code that uses functions in [zarr.storage](https://zarr.readthedocs.io/en/stable/api/storage.html) that no longer exist in v3
- Explaining what the alternative to [zarr.n5](https://zarr.readthedocs.io/en/stable/api/n5.html) is
- Explaining how to update code that imports from [zarr.codecs](https://github.com/zarr-developers/zarr-python/blob/support/v2/zarr/codecs.py)
- Explaining how to update code that uses [zarr.attrs](https://zarr.readthedocs.io/en/stable/api/attrs.html)
- Explaining how to update code that uses [zarr.sync](https://zarr.readthedocs.io/en/stable/api/sync.html)

File: docs/user-guide/v3_migration.rst [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+~~~~~~~~~~~~~~~~~~~~~~
+
+Zarr-Python 3 is still under active development, and is not yet fully complete.
+The following list summarizes areas of the codebase that we expect to build out
+after the 3.0 release:
+
+- The following functions / methods have not been ported to Zarr-Python 3 yet:
+
+  * :func:`zarr.copy`
+  * :func:`zarr.copy_all`
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:177]: According to https://github.com/zarr-developers/zarr-python/issues/2407#issuecomment-2494045572, copy and copy_all aren't going to be ported at all?

**Comment by d-v-b** [comment]
In addition to a migration guide, what do people think about a "zarr v2 vs zarr v3" section that just describes how the semantics of v2 and v3 arrays are different? Not sure if that's in scope for this PR, or if it should be a separate effort, but it seems like the content is highly overlapping. I think the migration guide could link to the "v2 vs v3 concepts" section as needed when people have questions like "what's the logic behind with dimension separator handling zarr-python 3.xx"

File: docs/user-guide/v3_migration.rst (hunk 1 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:16]: nice suggestion. thanks.

File: docs/user-guide/v3_migration.rst (hunk 4 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:93]: It is no longer possible (in the same way as it was before). I've clarified this.

File: docs/user-guide/v3_migration.rst (hunk 5 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:105]: yes, they are drop in replacements.

File: docs/user-guide/v3_migration.rst (hunk 8 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:128]: done.

File: docs/user-guide/v3_migration.rst (hunk 7 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:118]: Added additional details for all stores. 

tldr; n5/abs stores have external alternatives. the rest would require a custom store that zarr-python is unlikely to maintain unless they are added to the spec.

File: docs/user-guide/v3_migration.rst (hunk 9 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:130]: done.

File: docs/user-guide/v3_migration.rst (hunk 10 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:141]: I disagree. The new runtime configuration object (`zarr.config`) collects a number of options that were present in v2 in other places.

File: docs/user-guide/v3_migration.rst (hunk 11 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:165]: added notes clarifying this.

File: docs/user-guide/v3_migration.rst (hunk 16 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:177]: Right, we haven't taken any action there yet though.

File: docs/user-guide/v3_migration.rst (hunk 6 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:116]: expanded the language on the async store interface

File: docs/user-guide/v3_migration.rst (hunk 3 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:30]: Reworked this section. I still think a "getting ready" section is useful because projects will choose when to migrate based on their own situation.

**Comment by jhamman** [comment]
I'm now reasonably happy with this as a first cut and suggest we do what we can to get this online ASAP. There are gaps that can be filled in as we go but this is much better than the current alternative (no migration guide).

---

to @d-v-b's https://github.com/zarr-developers/zarr-python/pull/2596#issuecomment-2567078867, let's add a zarr v2 vs v3 page elsewhere. I will not try to squeeze that into this PR.

to @dstansby's comment about specific migration paths for stuff in `zarr.hierarchy`, `zarr.attrs`, etc. I believe the statement in the docs reflects our position: 

> The following internal modules are being removed or significantly changed. If your application relies on imports from any of the below modules, you will need to either a) modify your application to no longer rely on these imports or b) vendor the parts of the specific modules that you need. 

I'm open to specific suggestions for how we could improve this wording.

File: docs/user-guide/v3_migration.rst [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+   Zarr specification. To continue using version 2, set ``zarr_format=2`` when creating arrays
+   or set ``default_zarr_version=2`` in Zarr's :ref:`runtime configuration <config>`.
+
+The Group class
+~~~~~~~~~~~~~~~
+
+1. Disallow direct construction - use :func:`zarr.open_group` or :func:`zarr.create_group`
+   instead of directly constructing the :class:`zarr.Group` class.
+2. Most of the h5py compatibility methods are deprecated and will issue warnings if used.
+   The following migration is suggested:
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:115]:
```suggestion
2. Most of the h5py compatibility methods are deprecated and will issue warnings if used.
   The following functions are drop in replacements that have the same signature and functionality:
```

File: docs/user-guide/v3_migration.rst [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+2. Switched to an asynchronous interface for all store methods that result in IO. This
+   change ensures that all store methods are non-blocking and are as performant as
+   possible.
+
+Beyond the changes store interface, a number of deprecated stores were also removed in Zarr-Python 3.
+See :issue:`1274` for more details on the removal of these stores.
+
+- ``N5Store`` - see https://github.com/zarr-developers/n5py for an alternative interface to N5 formatted data.
+- ``ABSStore`` - use the :class:`zarr.storage.FsspecStore` instead along with fsspec's `adlfs backend <https://github.com/fsspec/adlfs>`_.
+- ``DBMStore``
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:137]:
```suggestion
The following stores have been removed altogether. Users who need these stores
will have to implement their own version in zarr-python v3.

- ``DBMStore``
```

File: docs/user-guide/v3_migration.rst (hunk 10 above)
Comment (dstansby) [reply docs/user-guide/v3_migration.rst:163]: Ah I didn't realise - in that case, this should be expanded to explain how users migrate from the old zarr.config system to the new system.

File: docs/user-guide/v3_migration.rst [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+after the 3.0 release. If features listed below are important to your use case of
+Zarr-Python, please open (or comment on) a
+`GitHub issue <https://github.com/zarr-developers/zarr-python/issues/new>`_.
+
+- The following functions / methods have not been ported to Zarr-Python 3 yet:
+
+  * :func:`zarr.copy` (:issue:`2407`)
+  * :func:`zarr.copy_all` (:issue:`2407`)
+  * :func:`zarr.copy_store` (:issue:`2407`)
+  * :func:`zarr.Group.move` (:issue:`2108`)
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:205]: It's a bit confusing that these say they're not ported, but are linked to functions in the API docs (which would imply they are implemented). Do they need removing from the API docs?

File: docs/user-guide/v3_migration.rst [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+
+- The following options in the top-level API (i.e. creation and convenience functions in
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:207]:
```suggestion
- The following features (corresponding to function arguments to functions in
```

File: docs/user-guide/v3_migration.rst [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+  :mod:`zarr`) have not been ported to Zarr-Python 3 yet.
+
+  * ``cache_attrs``
+  * ``cache_metadata``
+  * ``chunk_store`` (:issue:`2495`)
+  * ``meta_array``
+  * ``object_codec`` (:issue:`2617`)
+  * ``synchronizer`` (:issue:`1596`)
+  * ``dimension_separator``
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:216]: All of these, apart from `synchronizer`, aree both present, and documented as working in the [API docs for `zarr.api.asynchronous.create()`](https://zarr.readthedocs.io/en/latest/_autoapi/zarr/api/asynchronous/index.html#zarr.api.asynchronous.create). Does this mean they have been ported, or the docstrings in the API docs need fixing? (or something else?)

File: docs/user-guide/v3_migration.rst (hunk 16 above)
Comment (dstansby) [reply docs/user-guide/v3_migration.rst]: Is the plan to eventually implement `copy` and `copy_all`, or not? Currently this PR implies yes, but https://github.com/zarr-developers/zarr-python/issues/2407#issuecomment-2494045572 implies no.

File: docs/user-guide/v3_migration.rst [hunk 22]
```diff
@@ -0,0 +1,226 @@
...
+major version update), but where possible we have maintained backwards compatibility
+in the most widely used parts of the API. This in the :class:`zarr.Array` and
+:class:`zarr.Group` classes and the "top-level API" (e.g. :func:`zarr.open_array` and
+:func:`zarr.open_group`).
+
+Getting ready for 3.0
+---------------------
+
+Before migrating to Zarr-Python 3, we suggest projects that depend on Zarr-Python take
+the following actions:
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:33]:
```suggestion
the following actions in order:
```

File: docs/user-guide/v3_migration.rst [hunk 23]
```diff
@@ -0,0 +1,226 @@
...
+        from numcodecs import Blosc
+        # instead of:
+        # from zarr import Blosc
+
+   - The ``zarr.v3_api_available`` feature flag is being removed. In Zarr-Python 3.0
+     the v3 API is always available, so you shouldn't need to use this flag.
+   - The following internal modules are being removed or significantly changed. If
+     your application relies on imports from any of the below modules, you will need
+     to either a) modify your application to no longer rely on these imports or b)
+     vendor the parts of the specific modules that you need.
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:55]: There's scope for much more descriptive and helpful info about what's changed/what's gone here. Specifically:

- `zarr.attrs` has moved to `zarr.core.attributes`
- `zarr.codecs` has gone, import from `numcodecs` instead
- `zarr.context` has gone, with no replacement
- `zarr.core.Array` has moved to `zarr.Array` (but see migration notes below on how the `Array` object has changed)
- `zarr.hierarchy.Group` has moved to `zarr.Group` (but see migration notes below on how the `Group` object has changed)
- `zarr.hierarchy.group` and `zarr.hierarchy.open_group` has moved to `zarr.group` and `zarr.open_group`
- `zarr.indexing` has moved to `zarr.core.indexing`
- `zarr.meta` has moved to `zarr.core.metadata.common`, `zarr.core.metadata.v2` and `zarr.core.metadata.v3`
- `zarr.storage` still exists, but <list of thinks that have changed or been deleted>
- `zarr.sync` has gone, with no replacement
- `zarr.util` has gone, with no replacement (since this was always intended to be internal helper code)
- `zarr.n5` has gone, see <link to other repo> for replacement

File: docs/user-guide/v3_migration.rst [hunk 24]
```diff
@@ -0,0 +1,226 @@
+3.0 Migration Guide
+===================
+
+Zarr-Python 3 represents a major refactor of the Zarr-Python codebase. Some of the
+goals motivating this refactor included:
+
+* adding support for the Zarr V3 specification (along with the Zarr V2 specification)
+* cleaning up internal and user facing APIs
+* improving performance (particularly in high latency storage environments like
+  cloud object store)
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:10]:
```suggestion
  cloud object stores)
```

File: docs/user-guide/v3_migration.rst [hunk 25]
```diff
@@ -0,0 +1,226 @@
+3.0 Migration Guide
+===================
+
+Zarr-Python 3 represents a major refactor of the Zarr-Python codebase. Some of the
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:4]:
```suggestion
Zarr-Python 3.0 represents a major refactor of the Zarr-Python codebase. Some of the
```
Or change 3.0 > 3 below - I'd have a soft preference for "3" over "3.0".

File: docs/user-guide/v3_migration.rst [hunk 26]
```diff
@@ -0,0 +1,226 @@
...
+goals motivating this refactor included:
+
+* adding support for the Zarr V3 specification (along with the Zarr V2 specification)
+* cleaning up internal and user facing APIs
+* improving performance (particularly in high latency storage environments like
+  cloud object store)
+
+To accommodate this, Zarr-Python 3.0 introduces a number of changes to the API, including a number
+of significant breaking changes and pending deprecations.
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:13]:
```suggestion
of significant breaking changes and deprecations.
```

File: docs/user-guide/v3_migration.rst (hunk 10 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:163]: there was not an "old zarr.config" system. Runtime parameters were scattered throughout the API.

File: docs/user-guide/v3_migration.rst [hunk 27]
```diff
@@ -0,0 +1,226 @@
...
+
+This page provides a guide highlighting the most notable changes to help you
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:15]:
```suggestion
This page provides a guide explaing breaking changes and deprecations to help you
```

File: docs/user-guide/v3_migration.rst (hunk 19 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:205]: All of these exist as stubs in the library but currently raise a `NotImplementedError`

File: docs/user-guide/v3_migration.rst (hunk 21 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:216]: We chose to not break the function signature but instead to raise either an error or warning if these options were provided. So yes, they exist in the signature, but they cannot be used.

File: docs/user-guide/v3_migration.rst [hunk 28]
```diff
@@ -0,0 +1,226 @@
...
+open a `GitHub issue <https://github.com/zarr-developers/zarr-python/issues/new>`_
+so we can improve this guide.
+
+Compatibility target
+--------------------
+
+The goals described above necessitated some breaking changes to the API (hence the
+major version update), but where possible we have maintained backwards compatibility
+in the most widely used parts of the API. This in the :class:`zarr.Array` and
+:class:`zarr.Group` classes and the "top-level API" (e.g. :func:`zarr.open_array` and
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:26]: There's a section lower down that explains how breaking changes have happened to `Array` and `Group`, which seems at odds to this statement that backwards compat has been maintained. I'd be pro removing the "Compatibility target" section altogether to avoid this contradiction?

File: docs/user-guide/v3_migration.rst [hunk 29]
```diff
@@ -0,0 +1,226 @@
...
+     * ``zarr.meta``
+     * ``zarr.meta_v1``
+     * ``zarr.storage``
+     * ``zarr.sync``
+     * ``zarr.types``
+     * ``zarr.util``
+     * ``zarr.n5``
+
+3. Test that your package works with v3. You can start testing against version 3 now
+   (pre-releases are being published to PyPI weekly).
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:72]:
```suggestion
3. Test that your package works with v3.
```

File: docs/user-guide/v3_migration.rst [hunk 30]
```diff
@@ -0,0 +1,229 @@
...
+.. note::
+   Development and maintenance of the 2.x release series has moved to the
+   `support/v2 <https://github.com/zarr-developers/zarr-python/tree/support/v2>`_ branch.
+   Issues and pull requests related to this branch are tagged with the
+   `V2 <https://github.com/zarr-developers/zarr-python/labels/V2>`_ label.
+
+Migrating to Zarr-Python 3
+--------------------------
+
+The following sections provide details on the most important changes in Zarr-Python 3.
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:96]:
```suggestion
The following sections provide details on breaking changes in Zarr-Python 3.
```

File: docs/user-guide/v3_migration.rst [hunk 31]
```diff
@@ -0,0 +1,229 @@
...
+2. Most of the h5py compatibility methods are deprecated and will issue warnings if used.
+   The following functions are drop in replacements that have the same signature and functionality:
+
+   - Use :func:`zarr.Group.create_array` in place of :func:`zarr.Group.create_dataset`
+   - Use :func:`zarr.Group.require_array` in place of :func:`zarr.Group.require_dataset`
+
+The Store class
+~~~~~~~~~~~~~~~
+
+Some of the biggest changes in Zarr-Python 3 are found in the ``Store`` class. The most
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:123]:
```suggestion
The most
```

File: docs/user-guide/v3_migration.rst (hunk 10 above)
Comment (dstansby) [reply docs/user-guide/v3_migration.rst:166]: Is there an example of code that someone would write in v2, which they now have to use the configuration system for instead? I think should be the measure of if something should go in the migration guide or not.

File: docs/user-guide/v3_migration.rst (hunk 19 above)
Comment (dstansby) [reply docs/user-guide/v3_migration.rst:208]: Should we remove them from the API docs then, so the docs don't give a misleading impression they exist?

File: docs/user-guide/v3_migration.rst (hunk 21 above)
Comment (dstansby) [reply docs/user-guide/v3_migration.rst:219]: 👍 I will open an issue to fix the API docs for these then

**Review by dstansby — CHANGES_REQUESTED** [review]
🙌 this is looking great! I left another round of review comments, and a couple of more general things:

- Near the top, language should be changed from (e.g.) "will be" to "has been", since these are changes that have now happened.
- Since zarr.sync has gone with no replacement, there should be a short section on how to do process and thread syncrhonization in v3 (if it is possible - if not, just a short bit explaining it's not possible should be added)

File: docs/user-guide/v3_migration.rst (hunk 21 above)
Comment (dstansby) [reply docs/user-guide/v3_migration.rst:219]: Would be good to alkso note on this page that they exist in function signatures, but raise errors or warnings.

**Comment by jhamman** [comment]
I'm going to wrap this up today. Fixing conflicts and resolving the final comments.

File: docs/user-guide/v3_migration.rst (hunk 25 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:4]: changed to 3 in most places except where specifically referring to the 3.0 release.

File: docs/user-guide/v3_migration.rst (hunk 21 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:218]: done.

**Review by dstansby — CHANGES_REQUESTED** [review]
These previous comments still need resolving somehow

- https://github.com/zarr-developers/zarr-python/pull/2596/files#r1900951361
- https://github.com/zarr-developers/zarr-python/pull/2596/files#r1900946457
- https://github.com/zarr-developers/zarr-python/pull/2596/files#r1900958061
- https://github.com/zarr-developers/zarr-python/pull/2596/files#r1900919883

File: docs/user-guide/v3_migration.rst (hunk 19 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:206]: No opinion on whether to keep them in the API docs.

File: docs/user-guide/v3_migration.rst (hunk 28 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:26]: A number of people asked for this explanation so I'm keen to keep it, if you have ideas for how to sharpen it, feel free to suggest additional language.

File: docs/user-guide/v3_migration.rst (hunk 23 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:56]: I did a version of this.

File: docs/user-guide/v3_migration.rst [hunk 32]
```diff
@@ -0,0 +1,226 @@
...
+so we can improve this guide.
+
+Compatibility target
+--------------------
+
+The goals described above necessitated some breaking changes to the API (hence the
+major version update), but where possible we have maintained backwards compatibility
+in the most widely used parts of the API. This in the :class:`zarr.Array` and
+:class:`zarr.Group` classes and the "top-level API" (e.g. :func:`zarr.open_array` and
+:func:`zarr.open_group`).
```
Comment (dstansby) [inline docs/user-guide/v3_migration.rst:27]:
```suggestion
The goals described above necessitated some breaking changes to the API (hence the
major version update), but where possible we have maintained backwards compatibility
with functions in the top-level ``zarr`` namespace (e.g. :func:`zarr.open_array` and :func:`zarr.open_group`).
```

File: docs/user-guide/v3_migration.rst (hunk 28 above)
Comment (dstansby) [reply docs/user-guide/v3_migration.rst:26]: 👍 I put a suggestion here: https://github.com/zarr-developers/zarr-python/pull/2596/files/9427b10cade59400810890237ff71a10118319ae#r1902114450

File: docs/user-guide/v3_migration.rst (hunk 10 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:168]: removed the config section.

File: docs/user-guide/v3_migration.rst (hunk 32 above)
Comment (jhamman) [reply docs/user-guide/v3_migration.rst:27]: We have also attempted to maintain backward compatibility for the Group/Array classes. Neither the top level API or the Array/Group classes hit 100% backward compatibility.

**Review by dcherian — APPROVED** [review]
LGTM. Didn't verify the correctness, just reviewed the text.
